Objects/typeobject.c

```diff
@@ -3398,15 +3398,64 @@ get_bases_tuple(PyObject *bases_in, PyType_Spec *spec)
     return PyTuple_Pack(1, bases_in);
 }
 
+static inline int
+check_basicsize_includes_size_and_offsets(PyTypeObject* type) {
+    if (type->tp_alloc != PyType_GenericAlloc) {
+        // Custom allocators can ignore tp_basicsize
+        return 1;
+    }
+
+    if (type->tp_base && type->tp_base->tp_basicsize > type->tp_basicsize) {
+        PyErr_Format(PyExc_TypeError,
+                    "tp_basicsize for type '%s' (%d) is too small for base '%s' (%d)",
+                    type->tp_name, type->tp_basicsize,
+                    type->tp_base->tp_name, type->tp_base->tp_basicsize);
+        return 0;
+    }
+    if (type->tp_weaklistoffset + (Py_ssize_t)sizeof(PyObject*) > (Py_ssize_t)type->tp_basicsize) {
+        PyErr_Format(PyExc_TypeError,
+                    "tp_basicsize for type '%s' (%d) is too small for weaklist offset %d",
```

Comment:
I wonder if _"weaklist offset is out of bounds"_ would be better? Ditto for the other offset error messages below.